cylc/flow/platform_lookup.py

```diff
@@ -0,0 +1,166 @@
+    +<---+ and 'localhost' in  |  +--+ host both match      |
+    |    | platform hosts?     |  |  +----------------------+
+         +---------------------+  |                 |No
+    |            |Yes             |  +--------------v-------+
+    |    +-------v--------------+ |  | batch system match   |
+    |    | RETURN this platform <-+--+ and regex of platform|
+    |    +----------------------+ Yes| name matches host    |
+    |                                +----------------------+
+    |                                  |No
+    +<---------------------------------+
+
+    Args:
+        job (dict):
+            Suite config [runtime][TASK][job] section
+        remote (dict):
+            Suite config [runtime][TASK][remote] section
+        platforms (dict):
+            Dictionary containing platfrom definitions.
+
+    Returns:
+        platfrom (str):
+            string representing a platform from the global config.
+
+    Raises:
+        PlatformLookupError:
+            If no matching platform can be a found an error is raised.
+
+    Example:
+        >>> platforms = {
+        ...         'desktop[0-9][0-9]|laptop[0-9][0-9]': {},
+        ...         'sugar': {
+        ...             'login hosts': 'localhost',
+        ...             'batch system': 'slurm'
+        ...         }
+        ... }
+        >>> job = {'batch system': 'slurm'}
+        >>> remote = {'host': 'sugar'}
+        >>> reverse_lookup(platforms, job, remote)
+        'sugar'
+        >>> remote = {}
+        >>> reverse_lookup(platforms, job, remote)
+        'localhost'
+    """
+    # These settings are removed from the incoming dictionaries for special
+    # handling later - we want more than a simple match:
+    #   - In the case of host we also want a regex match to the platform name
+    #   - In the case of batch system we want to match the name of the system
+    #     to a platform when host is localhost.
+    if 'host' in remote.keys():
+        task_host = remote.pop('host')
+    else:
+        task_host = 'localhost'
+    if 'batch system' in job.keys():
+        task_batch_system = job.pop('batch system')
+    else:
+        # Necessary? Perhaps not if batch system default is 'background'
+        task_batch_system = 'background'
+
+    # Riffle through the platforms looking for a match to our task settings.
+    # reverse dict order
+    platforms = {
+        key: platforms[key] for key in reversed(list(platforms.keys()))
+    }
+
+    for platform_name, platform_spec in platforms.items():
+        # Handle all the items requiring an exact match.
+        generic_items_match = True
+        for task_section in [job, remote]:
+            shared_items = set(
+                task_section.keys()).intersection(set(platform_spec.keys()))
+            for shared_item in shared_items:
+                # breakpoint()
+                if platform_spec[shared_item] != task_section[shared_item]:
+                    generic_items_match = False
```

Comment:
@oliver-sanders ' solution is nicest, but otherwise best break out of the loop here (as soon as non-match is detected).

Reply:
Trouble is that we really want to use `continue` on both this loop and the one above, so I've gone with Oliver's solution. It's still much nicer than mine, and I've been able to remove the initial `generic_items_match = True` assignment.